### ISSUE TITLE
Segfault on Uniform Update

### ISSUE THREAD

**Issue by DaFluffyPotato** [moderngl/moderngl#580]
**Description**

ModernGL randomly causes segfaults when setting uniforms. It happens *very* rarely. Originally I had it happen within 5 seconds of startup on ModernGL 5.6.4. After updating to 5.8.2, I thought the issue went away, but it just happened for the first time today. Nothing particularly unusual was happening when the segfault occurred, but this time it was long after the first 5 seconds.

Due to the rarity of the segfault, I'm unable to provide stripped down code that reproduces the issue. Unfortunately the error occurs in my several thousand line application. I can add logging to determine what uniforms were being used upon segfault and update the issue the next time it happens though.

Here's my full stack trace:
```
pygame-ce 2.2.0 (SDL 2.0.22, Python 3.8.6)
Fatal Python error: (pygame parachute) Segmentation Fault
Python runtime state: initialized
 
Current thread 0x00003f64 (most recent call first):
  File "D:\Python38\lib\site-packages\moderngl\program_members\uniform.py", line 167 in value
  File "c:\Users\cmlsc\Desktop\underground fishing pro\scripts\pygpen\mgl\render_object.py", line 34 in update
  File "c:\Users\cmlsc\Desktop\underground fishing pro\scripts\pygpen\mgl\render_object.py", line 50 in render
  File "c:\Users\cmlsc\Desktop\underground fishing pro\scripts\rendering\raycast.py", line 42 in render_raycast
  File "c:\Users\cmlsc\Desktop\underground fishing pro\ufp.py", line 101 in update
  File "c:\Users\cmlsc\Desktop\underground fishing pro\scripts\pygpen\misc\game.py", line 20 in run
  File "c:\Users\cmlsc\Desktop\underground fishing pro\ufp.py", line 116 in <module>
```

Here's the relevant line of code from the second line of the stacktrace:
```py
self.program[uniform].value = uniforms[uniform]
```
`self.program` is a program created by `self.e['MGL'].ctx.program(vertex_shader=vert_shader, fragment_shader=frag_shader)` (`Context.Program()`)

The stacktrace revealed which program the error occurred on, but not the uniforms being used at the time.

Here are the fragment shader uniforms available for assignment:
```
uniform sampler2D map;
uniform vec3 lights[max_lights];
uniform int light_count;
```

The vertex shader uses no uniforms.

**Version info**

```
moderngl 5.8.2
--------------
vendor: NVIDIA Corporation
renderer: NVIDIA GeForce GTX 1070 Ti/PCIe/SSE2
version: 3.3.0 NVIDIA 516.94
python: 3.8.6 (tags/v3.8.6:db45529, Sep 23 2020, 15:52:53) [MSC v.1927 64 bit (AMD64)]
platform: win32
code: 330
```

**Comment by DaFluffyPotato** [comment]
just found out it was 5.6.4 and I had my installs mixed up.

**Comment by DaFluffyPotato** [comment]
Just confirmed it still happens on 5.8.2 today (shortly after startup for me). This time it was on a different project using the same framework. Happened multiple times to someone on mac while not following the pattern of occurring shortly after startup. Here's the error they sent me:
```
Fatal Python error: pygame_parachute: (pygame parachute) Segmentation Fault
Python runtime state: initialized

Current thread 0x0000000205465e00 (most recent call first):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/_moderngl.py", line 93 in write
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/_moderngl.py", line 79 in value
  File "/Users/user/Desktop/Of_m_and_m_test/milestone_2.8.5/murder-and-moonshine-main/scripts/pygpen/mgl/render_object.py", line 33 in update
  File "/Users/user/Desktop/Of_m_and_m_test/milestone_2.8.5/murder-and-moonshine-main/scripts/pygpen/mgl/render_object.py", line 48 in render
  File "/Users/user/Desktop/Of_m_and_m_test/milestone_2.8.5/murder-and-moonshine-main/scripts/rendering/raycast.py", line 49 in render_raycast
  File "/Users/user/Desktop/Of_m_and_m_test/milestone_2.8.5/murder-and-moonshine-main/mam.py", line 94 in update
  File "/Users/user/Desktop/Of_m_and_m_test/milestone_2.8.5/murder-and-moonshine-main/scripts/pygpen/misc/game.py", line 20 in run
  File "/Users/user/Desktop/Of_m_and_m_test/milestone_2.8.5/murder-and-moonshine-main/mam.py", line 125 in <module>
  ```
Line 33 is the same line of code mentioned as causing error in the initial issue report. It's basically the same shader as well (something for raycasted lighting).

**Comment by szabolcsdombi** [comment]
updating uniform by value (inf, float, tuple, list of ...) is already packed on python side, so there should be no segfault around parsing, see,

https://github.com/moderngl/moderngl/blob/897d5cbd9755dbea51ce772c04c3089dbc5eba1c/_moderngl.py#L71-L81

your last stack trace leads to:

https://github.com/moderngl/moderngl/blob/897d5cbd9755dbea51ce772c04c3089dbc5eba1c/src/moderngl.cpp#L8619-L8702

and we clearly have no size check on the raw data.

**Comment by DaFluffyPotato** [comment]
So a uniform of some array type is being written to with source data of an invalid length I'm guessing? I've added more logging to catch the exact uniform it happens on if it happens again. Also, Einarf recommended mentioning that the time it happened on Mac was on an M2 chip, although I'd assume that's unrelated at this point.

**Comment by DaFluffyPotato** [comment]
I'm guessing the fix known already, but just in case it's helpful, it just happened again and I caught the exact uniform. It's `uniform vec3 lights[max_lights];` with the assignment value being `[[0, 0, 0]]` and `max_lights` being `64`. Apparently it was an edge case in my code where `[[0, 0, 0]]` could be used instead of `[[0, 0, 0]] * 64`, although most of the time, the edge case worked as expected even with `[[0, 0, 0]]` being used and didn't cause any errors. It seems to just randomly (and rarely) cause ModernGL to segfault.

**Comment by szabolcsdombi** [comment]
setting shorter arrays should work.

a fix is required to change:

https://github.com/moderngl/moderngl/blob/897d5cbd9755dbea51ce772c04c3089dbc5eba1c/_moderngl.py#L97

`self.array_length` should be replaced with `len(data) // self.element_size` or similar.

**Comment by szabolcsdombi** [comment]
actually for moderngl I think shorter arrays were never expected. getter setter for uniforms were expecting full data. it should have thrown an error at least. the previous uniform value parser were doing the checks. this newer code is lacking it.

**Comment by szabolcsdombi** [comment]
fixed in #021f47f1
